design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,483 @@
+window. Some users with very large k8s clusters might only drain/reboot a portion of their nodes
+during a maintenance window over a period of several days. (Service ticket TTL could be extended,
+but extending it to several days would also have Ceph security implications that are best avoided
+if possible)
+
+In cases where the rotated key can't be picked up quickly, the alternative is for Rook to create a
+new ceph client (user) with the new key without deleting the old user/key. Then, Rook updates the
+CSI secrets to have the new user+key. This would begin the window in which users can drain/reboot
+nodes. Any new PVC mounts would use the new key, and old mounts would continue operating happily
+since the old user+key are not yet destroyed.
+
+After the user is done with node updates, the user should have some way of indicating to Rook that
+the old key is no longer needed, telling Rook it is safe to delete the old key. Autodetection might
+be possible, but let's leave that for a future exercise.
+
+For clarity, this design could be called "overlapping" rotation, and the design elsewhere in this
+doc "in-place" rotation.
+
+Overlapping rotation will be implemented for 3 key types:
+
+- CSI keys
+- RBD mirror peer keys (not daemon keys)
+- CephFS mirror peer (not daemon keys)
+
+## User interfaces
+
+### Key rotation base interface
+
+```yaml
+keyRotationPolicy: Disabled | WithCephVersionUpgrade | KeyGeneration
+keyGeneration: <int> # used with KeyGeneration
+keepPriorKeyCount: <int> # only present to configure overlapping rotation
+```
+
+- `keyRotationPolicy` (string): select a key rotation policy:
+    - `Disabled` (default when unset): don't rotate keys after initial creation
+    - `WithCephVersionUpgrade`: rotate keys when the Ceph version updates
+    - `KeyGeneration`: rotate when the `keyGeneration` input is greater than the current key generation
+
+- `keyGeneration` (int): when `keyRotationPolicy==KeyGeneration`, set this to the desired key
+    generation value. Ignored for other rotation policies. If this is set to greater than the
+    current key generation, relevant keys will be rotated, and the generation value will be updated
+    to this new value (generation values are not necessarily incremental, though that is the
+    intended use case).
+    If this is set to less than or equal to the current key generation, keys are not rotated.
+
+API for overlapping rotation:
+
+- `keepPriorKeyCount` (int): only available for components that use overlapping key rotation. This
+    tells Rook how many prior keys to keep active. Generally, this would be set to `1` to allow for
+    a migration period for applications. If desired, set this to `0` to delete prior keys after
+    migration.
+
+### Enabling daemon key rotations
+
+Ceph developers have suggested rotation on every Ceph version update. This corresponds to a rotation
+each time the administrator updates the CephCluster's Ceph image. This also aligns with when Rook
```

Comment:
Just to clarify if rook base image is ahead of actual ceph version, do we still need rotation?